optuna/samplers/_nsga2.py

```diff
@@ -327,3 +395,52 @@ def _crowding_distance_sort(population: List[FrozenTrial]) -> None:
 
     population.sort(key=lambda x: manhattan_distances[x.number])
     population.reverse()
+
+
+def _constrained_dominates(
+    trial0: FrozenTrial, trial1: FrozenTrial, directions: Sequence[StudyDirection]
+) -> bool:
+    """Checks constrained-domination.
+
+    A trial x is said to constrained-dominate a trial y, if any of the following conditions is
+    true:
+    1) Trial x is feasible and trial y is not.
+    2) Trial x and y are both infeasible, but solution x has a smaller overall constraint
+    violation.
+    3) Trial x and y are feasible and trial x dominates trial y.
+    """
+
+    constraints0 = trial0.system_attrs[_CONSTRAINTS_KEY]
+    constraints1 = trial1.system_attrs[_CONSTRAINTS_KEY]
+
+    assert isinstance(constraints0, (list, tuple))
+    assert isinstance(constraints1, (list, tuple))
+
+    if len(constraints0) != len(constraints1):
+        raise ValueError("Trials with different numbers of constraints cannot be compared.")
+
+    if trial0.state != TrialState.COMPLETE:
+        return False
+
+    if trial1.state != TrialState.COMPLETE:
+        return True
+
+    satisfy_constraints0 = all(v <= 0 for v in constraints0)
+    satisfy_constraints1 = all(v <= 0 for v in constraints1)
+
+    if satisfy_constraints0 and satisfy_constraints1:
+        # Both trials satisfy the constraints.
+        return _dominates(trial0, trial1, directions)
+
+    if satisfy_constraints0:
+        # trial0 satisfies the constraints, but trial1 violates them.
+        return True
+
+    if satisfy_constraints1:
+        # trial1 satisfies the constraints, but trial0 violates them.
+        return False
+
+    # Both trials violate the constraints.
+    violation0 = sum(v for v in constraints0 if v > 0)
+    violation1 = sum(v for v in constraints1 if v > 0)
+    return violation0 < violation1
```

Comment:
Would this be more aligned with how [`_dominates` is implemented](https://github.com/optuna/optuna/blob/master/optuna/_multi_objective.py#L58) ?
```suggestion
    return violation0 <= violation1
```

Reply:
Thank you for your suggestion, but I guess the current code follows the original `_dominates` function because both of them return `False` if x and y are identical.

https://github.com/optuna/optuna/blob/baf56206a357ffe0b1adb95289a6f4d3034c3ad3/optuna/_multi_objective.py#L55-L58

In addition,  the [original paper](https://www.iitk.ac.in/kangal/Deb_NSGA-II.pdf) excludes the case of `violation0 == violation1`:

> 2) Solutions and are both infeasible, but solution has a smaller overall constraint violation.

Reply:
Thanks, I must have missed the early return.